# Assignment: Data Wrangling
## `! git clone https://github.com/DS3001/wrangling`
## Do Q2, and one of Q1 or Q3.

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1. The paper is about data tidying, which is an important aspect regarding data cleaning. Considering the huge amount of time that most people spent on cleaning data, the author points out the necessity to have a consistant data structure corresponding with minimal nubmers of tools.
2. "tidy data standard" is intended to provide a standard, unified ways to organize data. Realizing the huge amount of time taken to clean the data, Wickham decides to provide a standard way to shorten the necessary time organizing the data and the effort needed to handle the connections between different steps.
3. The first sentence means that tidy datasets follow a consistent structure, making them easier to work with, while messy datasets are unpredictable and unique in their problems. The second sentence says that while it's usually easy to identify variables and observations in a specific dataset, it’s hard to define them universally because each dataset has its own context.
4. Wickham defines values as individual data points, variables as the things being measured, such as temperature, and observations as the set of values related to a single unit (like a person or day).
5. "Tidy data" is defined as each variable is a column, each observation is a row, and each type of observational unit gets its own table. 
6. Messy data often has column headers that should be values, multiple variables crammed into one column, variables spread out in both rows and columns, different types of data mixed into one table, or data about one unit split across multiple tables. Table 4 is messy because the income categories are stuck in the column headers instead of being treated as values. Melting fixes this by converting those columns into rows and making the dataset more workable.
7. Table 11 is a mess because it stores tmin and tmax as rows instead of columns, which breaks the tidy data structure. In contrast, Table 12 is tidy because it puts everything into proper columns, making it easier to analyze.
8. The "chicken-and-egg" issue is that tidy data is only as useful as the tools we have to work with it. Wickham hopes that as more tools are built around tidy data, the process of working with and cleaning data will get even easier, pushing the field forward.

**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [46]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

#Q2-1
df = pd.read_csv('./data/airbnb_hw.csv',low_memory=False)
df.head()
print(df.dtypes, '\n')
print(df["Price"].unique(),'\n')

num_price = df["Price"].str.replace(",", "")
num_price = pd.to_numeric(num_price, errors="coerce")
print("Total Missings:", sum(num_price.isnull()),"\n")

Host Id                         int64
Host Since                     object
Name                           object
Neighbourhood                  object
Property Type                  object
Review Scores Rating (bin)    float64
Room Type                      object
Zipcode                       float64
Beds                          float64
Number of Records               int64
Number Of Reviews               int64
Price                          object
Review Scores Rating          float64
dtype: object 

['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1,990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' 

I start with using "dtype" to check the types of the variables and I found out that the type of Price is object instead of integer. Then I looked up the values of elements in Price using "unique" and I found out that the type of Price is string and there is a comma seperator for the thousands, which stops me from directly using "to_numeric" to transfer the type from string to integer. Therefore, I used "str.replace" to remove the comma seperator first and then did the transformation. Finally, I used "isnull" to sum all the null values in order to find the nubmer of missing values. The output showed no value is missing.

In [49]:
#Q2-2
df = pd.read_csv('./data/sharks.csv',low_memory=False)
df.head()
print(df.dtypes, '\n')
print(df["Type"].unique(),'\n')
type = df["Type"]
type = df["Type"].replace(["Questionable", "Unconfirmed", "Unverified", "Invalid", "Under investigation"], np.nan)
type = type.replace(["Watercraft", "Boating", "Sea Disaster", "Boat", "Boatomg"],"Watercraft_Related")
df["Type"] = type
print(df["Type"].unique(),'\n')
del type

index             int64
Case Number      object
Date             object
Year            float64
Type             object
                 ...   
Unnamed: 251    float64
Unnamed: 252    float64
Unnamed: 253    float64
Unnamed: 254    float64
Unnamed: 255    float64
Length: 257, dtype: object 

['Unprovoked' 'Provoked' 'Questionable' 'Watercraft' 'Unconfirmed'
 'Unverified' 'Invalid' 'Under investigation' 'Boating' 'Sea Disaster' nan
 'Boat' 'Boatomg'] 

['Unprovoked' 'Provoked' nan 'Watercraft_Related'] 



I combine ["Questionable", "Unconfirmed", "Unverified", "Invalid", "Under investigation"] together as  nan because those are all uncertain situation. Then I combine ["Watercraft", "Boating", "Sea Disaster", "Boat", "Boatomg"] to "Watercraft_Related" because these are all the types related to watercraft. After the cleaning process, I have 4 vairables for the type, which are ['Unprovoked' 'Provoked' nan 'Watercraft_Related'] 

In [51]:
#Q2-3
url = 'http://www.vcsc.virginia.gov/pretrialdataproject/October%202017%20Cohort_Virginia%20Pretrial%20Data%20Project_Deidentified%20FINAL%20Update_10272021.csv'
df = pd.read_csv(url,low_memory=False)
de_release = df['WhetherDefendantWasReleasedPretrial']
print("Before cleaning: ", de_release.unique(),'\n')
de_release = de_release.replace(9,np.nan)
print("After cleaning: ", de_release.unique(),'\n')
df['WhetherDefendantWasReleasedPretrial'] = de_release
del de_release

Before cleaning:  [9 0 1] 

After cleaning:  [nan  0.  1.] 



In [67]:
#Q2-4
length = df["ImposedSentenceAllChargeInContactEvent"]
type = df["SentenceTypeAllChargesAtConvictionInContactEvent"]

# print(length.unique(),"\n")
length = pd.to_numeric(length, errors="coerce")
print("Number of nan value Before: ", sum(length.isnull()), "\n")
lengthisnull = length.isnull()
print(pd.crosstab(lengthisnull,type),"\n")

length = length.mask(type == 4, 0)

print("Number of nan value After: ", sum(length.isnull()), "\n")

df['ImposedSentenceAllChargeInContactEvent'] = length

del length, type

Number of nan number Before:  9053 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274 

Number of nan number After:  274 



I start with finding the type of "ImposedSentenceAllChargeInContactEvent" and I found that the type is string instead of float, so the first step I did is to change the type from string to numerical value using "pd.to_numeric". Then I applied sum to find out the number of nan number and realized that result if not ideal due to the hugh nan number. Given the hint I used "crosstab" to find the relationship between "SentenceTypeAllChargesAtConvictionInContactEvent" and "ImposedSentenceAllChargeInContactEvent". I found that most nan values occurs when "SentenceTypeAllChargesAtConvictionInContactEvent" is 4. Therefore, I masked the value based on the condition ("SentenceTypeAllChargesAtConvictionInContactEvent" = 4). The number of nan value after reduces to 274.

**Q3.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?